riscv/src/register/macros.rs

```diff
@@ -397,3 +397,446 @@ macro_rules! clear_pmp {
         }
     };
 }
+
+/// Helper macro to create a read-write CSR type.
+///
+/// The type allows to read the CSR value into memory, and update the field values in-memory.
+///
+/// The user can then write the entire bitfield value back into the CSR with a single write.
+#[macro_export]
+macro_rules! read_write_csr {
+    ($doc:expr,
+     $ty:ident: $csr:tt,
+     $mask:tt,
+     $(
+     $field_doc:expr,
+     $field:ident: $get_bit:literal,
+     $set_field_doc:expr,
+     $set_field:ident: $set_bit:literal$(,)?
+     )+) => {
+        $crate::read_write_csr!($doc, $ty: $csr, $mask);
+
+        $(
+            $crate::read_write_csr_field!(
+                $ty,
+                $field_doc,
+                $field: $get_bit,
+                $set_field_doc,
+                $set_field: $set_bit,
+            );
+        )+
+    };
```

Comment:
> Less repetition in macros. Maybe we could atomize them and leave one "core" macro that creates the struct with bitmask and from_bits/bits, another for the get/set...

That's essentially what is there, each of the branches for the various ways to call `[read,write]_[only,write]_csr` call the internal branch for creating the struct.

Then, the repeated part (in the `$(...)+`) creates the getters/setters. I guess we could also make `read_write_csr_field` into another branch of the macro...

Other than that, I'm not sure exactly what you're asking for. What changes would you like to see?

Reply:
What I mean is that, for example, [this](https://github.com/rust-embedded/riscv/blob/fd163b109404372983ef0fa82fec518d9ccac2d4/riscv/src/register/macros.rs#L571) part of the macro where the register struct is defined appears three times. It would be nice to have it somehow just once in a separate, basic macro and, depending on the macro branches, call this macro

Reply:
Ah, I see what you mean. Because each of the macros has this helper branch, make it into its own macro that each one could call.

Alright, I'll work on separating it out. Maybe with a note that you can, but probably don't want to, call this helper macro on its own.